producer/writer/consumer_writer.go

```diff
@@ -0,0 +1,317 @@
+	decodeLock sync.Mutex
+	encdec     proto.EncodeDecoder
+
+	addr            string
+	retrier         retry.Retrier
+	router          ackRouter
+	opts            Options
+	resetDelayNanos int64
+	logger          log.Logger
+
+	lastResetNanos int64
+	closed         bool
+	doneCh         chan struct{}
+	resetCh        chan struct{}
+	closeWG        sync.WaitGroup
+	m              consumerWriterMetrics
+
+	nowFn     clock.NowFn
+	connectFn connectFn
+}
+
+// TODO: Remove the nolint comment after adding usage of this function.
+// nolint: deadcode
+func newConsumerWriter(
+	addr string,
+	router ackRouter,
+	opts Options,
+) consumerWriter {
+	if opts == nil {
+		opts = NewOptions()
+	}
+	w := &consumerWriterImpl{
+		encdec:          proto.NewEncodeDecoder(defaultConn, opts.EncodeDecoderOptions()),
+		addr:            addr,
+		retrier:         retry.NewRetrier(opts.ConnectionRetryOptions().SetForever(defaultRetryForever)),
+		router:          router,
+		opts:            opts,
+		resetDelayNanos: int64(opts.ConnectionResetDelay()),
+		logger:          opts.InstrumentOptions().Logger(),
+		lastResetNanos:  0,
+		closed:          false,
+		doneCh:          make(chan struct{}),
+		resetCh:         make(chan struct{}, 1),
+		m:               newConsumerWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+		nowFn:           time.Now,
+	}
+
+	w.connectFn = w.connectOnce
+	if err := w.resetWithConnectFn(w.connectFn); err != nil {
+		w.signalReset()
+	}
+	return w
+}
+
+// Write should fail fast so that the write could be tried on other
+// consumer writers that are sharing the data.
+func (w *consumerWriterImpl) Write(m proto.Marshaler) error {
+	w.encodeLock.Lock()
+	err := w.encdec.Encode(m)
+	w.encodeLock.Unlock()
+	if err != nil {
+		w.m.encodeError.Inc(1)
+		w.signalReset()
+	}
+	return err
+}
+
+func (w *consumerWriterImpl) signalReset() {
+	w.RLock()
+	lastResetNanos := w.lastResetNanos
+	w.RUnlock()
+	// Avoid resetting too frequent.
+	if w.nowFn().UnixNano() < lastResetNanos+w.resetDelayNanos {
+		return
+	}
+	select {
+	case w.resetCh <- struct{}{}:
+	default:
+	}
+}
+
+func (w *consumerWriterImpl) Init() {
+	w.closeWG.Add(2)
+	go w.readAcksForever()
+	go w.resetConnectionForever()
+}
+
+func (w *consumerWriterImpl) readAcksForever() {
+	defer w.closeWG.Done()
+
+	var acks msgpb.Ack
+	for !w.isClosed() {
+		w.decodeLock.Lock()
+		err := w.encdec.Decode(&acks)
+		w.decodeLock.Unlock()
+		if err != nil {
+			w.m.decodeError.Inc(1)
+			w.signalReset()
+			// Adding some delay here to avoid this being retried in a tight loop
+			// when underlying connection is misbehaving.
+			time.Sleep(w.opts.AckErrorRetryDelay())
+			continue
+		}
+		for _, m := range acks.Metadata {
+			if err := w.router.Ack(metadataFromProto(m)); err != nil {
+				w.m.ackError.Inc(1)
+				w.logger.Errorf("could not ack metadata, %v", err)
+			}
+		}
+		// NB(cw) The proto needs to be cleaned up because the gogo protobuf
```

Comment:
%s/gogo/go ?

Reply:
we are using gogo protobuf: https://github.com/gogo/protobuf which is different than go protobuf